### ISSUE TITLE
Why only use the first patch? Thanks

### ISSUE THREAD

**Issue by junyongyou** [lucidrains/vit-pytorch#18]
I don't understand in line 124 of vit_pytorch.py:
`x = self.to_cls_token(x[:, 0])`
If the first dimension of x is batch, then the 2nd dimension 0 should be patch, as the dimension of x should be [batch, patch, feature]. Does it mean only the first patch is used? Could anybody help me on this? Thanks a lot.

**Comment by lucidrains** [comment]
@junyongyou this is just a classic technique from historical usage of attention networks in NLP, from the very first BERT paper

the idea is basically to provide a token, that after each subsequent layer, aggregates information from the rest of the tokens without being obstructed, thus named the cls token (classification token)

however, i've seen other attention nets work just as well even without cls token, meaning or summing the final row of embeddings. meanwhile, other papers are investigating > 1 cls token (https://arxiv.org/pdf/2006.11527.pdf), viewing them as a form of memory. it's not too critical for ViT either way

**Comment by lucidrains** [comment]
@junyongyou ohh, i understand your confusion now. so it's actually not the first patch, but the cls token concatted here https://github.com/lucidrains/vit-pytorch/blob/main/vit_pytorch/vit_pytorch.py#L118

**Comment by junyongyou** [comment]
> 
> 
> @junyongyou ohh, i understand your confusion now. so it's actually not the first patch, but the cls token concatted here https://github.com/lucidrains/vit-pytorch/blob/main/vit_pytorch/vit_pytorch.py#L118

Thank you very much @lucidrains I am not familiar with NLP, and PyTorch either. I looked at some TensorFlow implementation of ViT, basically they both follow your PyTorch implementation. I understand x is concatenated with cls token before x = self.to_cls_token(x[:, 0]). But I still suspect some information might be lost here, if only the first column of the 2nd dimension is used, but I am not sure. I have looked at the TF tutorial of [Transformer](https://www.tensorflow.org/tutorials/text/transformer). It seems this approach is not used there. But of course the decoder is used in language modelling. Anyways, thank you very much for the explanation.
